### PR TITLE
Add non_blocking to Tensor/Module.to

### PR DESCRIPTION
**Pull request by ssnl**
cc @gchanan

### CODE REVIEW

File: torch/nn/modules/module.py [hunk 1]
```diff
@@ -353,44 +356,83 @@ def arg_error():
...
-            for arg in args:
-                if isinstance(arg, torch.dtype):
-                    if dtype is not None:
-                        raise arg_error()
-                    dtype = arg
-                else:
-                    if device is not None:
-                        raise arg_error()
-                    device = arg
+            # 'device' has priority because of .to(device, dtype=None, non_blocking=False)
```
Comment (ezyang) [inline torch/nn/modules/module.py]: Aww man, dis argument parsing code ^^

Comment (SsnL) [reply torch/nn/modules/module.py]: yeah I know :/

File: test/test_nn.py [hunk 2]
```diff
@@ -1183,15 +1183,15 @@ def test_type(self):
...
+            net.to("cuda", torch.double, True)
             self.assertIsInstance(l.weight.data, torch.cuda.DoubleTensor)
             self.assertIsInstance(l.bias.data, torch.cuda.DoubleTensor)
             self.assertIsInstance(net.indices, torch.cuda.LongTensor)
             net.to(device="cuda:0", dtype=torch.half)
             self.assertIsInstance(l.weight.data, torch.cuda.HalfTensor)
             self.assertIsInstance(l.bias.data, torch.cuda.HalfTensor)
             self.assertIsInstance(net.indices, torch.cuda.LongTensor)
-        net.to(torch.device("cpu"))
+        net.to(torch.device("cpu"), non_blocking=True)
```
Comment (apaszke) [inline test/test_nn.py]: Considering how complicated the parsing code gets, we should also start getting negative tests (i.e. make sure it fails for bad arg configurations)

**Review by apaszke — COMMENTED** [review]
(Not a complete review, I only have a suggestion)

**Review by colesbury — COMMENTED** [review]
Why is Module.to() so complicated?

Why isn't it something like:

```
def to(self, device, dtype=None, non_blocking=False)
  if isinstance(device, torch.dtype):
    if dtype is not None:
       # raise duplicate dtype error message
    dtype = device
    device = None
  if device is not None:
    # validate device
  if dtype is not None:
    # validate dtype

  self.apply_(lambda t: t.to(device, dtype, non_blocking))
```

File: torch/nn/modules/module.py [hunk 3]
```diff
@@ -296,17 +296,20 @@ def to(self, *args, **kwargs):
 
         This can be called as
 
-        .. function:: to(device)
-
         .. function:: to(dtype)
 
-        .. function:: to(device, dtype)
+        .. function:: to(device, dtype=None, non_blocking=True)
```
Comment (colesbury) [inline torch/nn/modules/module.py]: Should this match `Tensor.to` which defaults to `non_blocking=False`?

Comment (SsnL) [reply torch/nn/modules/module.py]: Good catch!

**Comment by SsnL** [comment]
@colesbury I didn't want the signature commit to one overload over another in the previous version. But I agree that now it is quite complicated. I'll change it.

File: torch/nn/modules/module.py (hunk 1 above)
Comment (gchanan) [reply torch/nn/modules/module.py]: is it possible to move this to C++ so we can just use the same function signatures?  Maybe make an internal function or something that returns a tuple idk.

File: tools/autograd/templates/python_variable_methods.cpp [hunk 4]
```diff
@@ -559,20 +559,20 @@ static PyObject * THPVariable_to(PyObject* self, PyObject* args, PyObject* kwarg
 {
   HANDLE_TH_ERRORS
   static PythonArgParser parser({
-    "to(Device device, ScalarType dtype=None)",
+    "to(Device device, ScalarType dtype=None, bool non_blocking=False)",
     "to(ScalarType dtype)",
```
Comment (gchanan) [inline tools/autograd/templates/python_variable_methods.cpp]: nit: I'd just throw non_blocking on here too even though it doesn't do anything (a lot of combinations don't do anything already).  It just makes it easier to write functions that take devices and/or dtypes without having to move the error checking up a level.

File: tools/autograd/templates/python_nn_functions.cpp [hunk 5]
```diff
@@ -15,9 +17,42 @@ using namespace torch::autograd::utils;
...
+    PyTuple_SET_ITEM(tuple.get(), 0, Py_None);
+  }
+  if (scalarType) {
+    PyTuple_SET_ITEM(tuple.get(), 1, torch::autograd::utils::wrap(torch::getDtype(*scalarType)));
+  } else {
+    Py_INCREF(Py_None);
+    PyTuple_SET_ITEM(tuple.get(), 1, Py_None);
+  }
+  if (non_blocking) {
+    Py_INCREF(Py_True);
```
Comment (gchanan) [inline tools/autograd/templates/python_nn_functions.cpp]: nit: you could collapse this case using utils::wrap on true/false and get rid of the Py_INCREF.

File: test/test_nn.py [hunk 6]
```diff
@@ -1173,6 +1173,28 @@ def test_add_module(self):
         self.assertEqual(net.l, l3)
         self.assertRaises(TypeError, lambda: net.add_module('x', 'non-module'))
 
+    def test_module_to_argparse(self):
+        net = nn.Sequential(nn.Linear(3, 3))
+        cpu = torch.device('cpu')
+        with self.assertRaises(TypeError):
+            net.to(cpu, True)
```
Comment (gchanan) [inline test/test_nn.py]: why is this an error?

**Comment by SsnL** [comment]
@gchanan let me know what you think about this! :)

File: tools/autograd/templates/python_variable_methods.cpp [hunk 7]
```diff
@@ -558,31 +559,25 @@ static PyObject * THPVariable_storage_type(PyObject* self, PyObject* arg)
...
-    return THPVariable_Wrap(torch::utils::dispatch_type_conversion(self_, type, deviceAutoGPU, false));
+  } else if (device) {
+    // device and maybe dtype given
+    auto& self_ = reinterpret_cast<THPVariable*>(self)->cdata;
+    auto deviceAutoGPU = device->deviceInt64();
+    auto& layout = *torch::getLayout(self_.type().backend());
+    auto& type = torch::getType(scalarType.value_or(self_.type().scalarType()), layout, device->type);
+    return THPVariable_Wrap(torch::utils::dispatch_type_conversion(self_, type, deviceAutoGPU, non_blocking));
+  } else {
+    Py_INCREF(self);
```
Comment (gchanan) [inline tools/autograd/templates/python_variable_methods.cpp]: nit: can you fold this case into the !device case (so there is no !device && scalarType case)?  type should be the same as in the below case, scalarType.value_or(self_.type().scalarType()), right?

**Review by gchanan — APPROVED** [review]

**Pull request merged**